### PR TITLE
Add sub roles to MacOS adapter

### LINKED ISSUE #372: macOS: Tabs are currently announced as radio buttons
This probably means that the mapping from the AccessKit role to the macOS (AppKit) role is incorrect or incomplete.

### LINKED ISSUE #477: MacOS Adapter - Sub roles
I have been trying to implement accessibility for some of my controls, and is currently comparing the accessiblity tree generated from my application, and from the equivalent SwiftUI application. 

When implementing accessibility for checkboxes, toggle buttons and switches, i noticed the accessibility tree generated by SwiftUI uses the role CheckBox for all the three controls, but with different sub roles.

https://developer.apple.com/documentation/appkit/nsaccessibilitysubrole?language=objc

How is it intended to implement these using AccessKit? I see AccessKit specifies CheckBox and Switch as different roles, but I see no role equivalent to the toggle button.

I can also see when the Switch and CheckBox roles from AccessKit are used, they both set the NSAccessibilityRole CheckBox, but no sub role.

## Toggle button from SwiftUI

Screenshot 2024-10-29 at 21 16 28

## Checkbox from SwiftUI

Screenshot 2024-10-29 at 21 17 13

## Toggle from SwiftUI

Screenshot 2024-10-29 at 21 18 04

### PR DESCRIPTION
**Pull request by HolgerGottChristensen**
Adds initial support for accessibility sub roles to the MacOS adapter, allowing AccessKit to expose Toggle buttons, Switch and more, correctly to VoiceOver.

Needed clarifications:
- [x] Does AccessKit handle math, and all the math roles from chromium? (Not currently available in AccessKit)
- [x] How are search boxes represented? [Role from MacOS](https://developer.apple.com/documentation/appkit/nsaccessibilitysearchfieldsubrole?language=objc) [Multiple roles from chromium](https://github.com/chromium/chromium/blob/92272b8fbd80b95c7539e7853b13fbe726b035f9/ui/accessibility/platform/ax_platform_node_cocoa.mm#L110C7-L111C55) (`Role::SearchInput`)
- [x] How are secure textfields represented? [Role from MacOS](https://developer.apple.com/documentation/appkit/nsaccessibilitysecuretextfieldsubrole?language=objc) (`Role::PasswordInput`)

Fixes #477 

Fixes #372

### CODE REVIEW

**Comment by DataTriny** [comment]
> How is search box represented?

By giving `Role::SearchInput` to a node.

> How is secure textfields represented?

By giving `Role::PasswordInput` to a node.

**Comment by HolgerGottChristensen** [comment]
I dont see any of the math roles in the list of roles: https://github.com/AccessKit/accesskit/blob/main/common/src/lib.rs#L55

Are they defined somewhere else?

**Comment by DataTriny** [comment]
Woops, sorry. We must have removed them at some point. Please do the same.

**Comment by DataTriny** [comment]
With the changes from this PR, tabs are correctly announced as tabs by VoiceOver. This can be tested by running the gallery example from [my fork of Slint](https://github.com/DataTriny/slint/tree/update-accesskit). I've updated the PR description accordingly.

**Comment by mwcampbell** [comment]
I don't recall dropping math roles. It's more likely that when I copied the roles from Chromium in July 2021, those roles didn't exist yet.

**Comment by HolgerGottChristensen** [comment]
> I don't recall dropping math roles. It's more likely that when I copied the roles from Chromium in July 2021, those roles didn't exist yet.

Yes, they were added in October 2021, so right after you copied the roles :)
https://github.com/chromium/chromium/commit/6ee95856677e69393c28b5c03142d29e79e69c49

File: platforms/macos/src/node.rs [hunk 1]
```diff
@@ -232,6 +232,65 @@ fn ns_role(node: &Node) -> &'static NSAccessibilityRole {
     }
 }
 
+fn ns_sub_role(node: &Node) -> &'static NSAccessibilitySubrole {
```
Comment (DataTriny) [inline platforms/macos/src/node.rs:235]:
```suggestion
fn sub_role(node: &Node) -> &'static NSAccessibilitySubrole {
```

File: platforms/macos/src/node.rs [hunk 2]
```diff
@@ -232,6 +232,65 @@ fn ns_role(node: &Node) -> &'static NSAccessibilityRole {
...
+            Role::AlertDialog => ns_string!("AXApplicationAlertDialog"),
+            Role::Article => ns_string!("AXDocumentArticle"),
+            Role::Banner => ns_string!("AXLandmarkBanner"),
+            Role::Button => {
+                if node.toggled().is_some() {
+                    NSAccessibilityToggleSubrole
+                } else {
+                    NSAccessibilityUnknownSubrole
+                }
+            }
```
Comment (DataTriny) [inline platforms/macos/src/node.rs:250]:
```suggestion
            Role::Button if node.toggled().is_some() => NSAccessibilityToggleSubrole,
```

File: platforms/macos/src/node.rs [hunk 3]
```diff
@@ -411,6 +470,13 @@ declare_class!(
                 .copy()
         }
 
+        #[method_id(accessibilitySubrole)]
+        fn sub_role(&self) -> Id<NSAccessibilitySubrole> {
+            self.resolve(ns_sub_role)
```
Comment (DataTriny) [inline platforms/macos/src/node.rs:475]:
```suggestion
            self.resolve(sub_role)
```

**Review by DataTriny — CHANGES_REQUESTED** [review]

**Review by DataTriny — APPROVED** [review]
Thank you @HolgerGottChristensen!

**Pull request merged**